apps/readiness_healthcheck.go

```diff
@@ -0,0 +1,133 @@
+	. "github.com/onsi/gomega/gexec"
+
+	"github.com/cloudfoundry/cf-acceptance-tests/helpers/random_name"
+	"github.com/cloudfoundry/cf-test-helpers/v2/cf"
+	"github.com/cloudfoundry/cf-test-helpers/v2/helpers"
+	"github.com/cloudfoundry/cf-test-helpers/v2/workflowhelpers"
+)
+
+var _ = AppsDescribe("Readiness Healthcheck", func() {
+	var appName, proxyAppName string
+	var orgName string
+	var spaceName string
+	var readinessHealthCheckTimeout = "25s" // 20s route emitter sync loop + 2s hc interval + bonus
+
+	BeforeEach(func() {
+		if !Config.GetReadinessHealthChecksEnabled() {
+			Skip(skip_messages.SkipReadinessHealthChecksMessage)
+		}
+		appName = random_name.CATSRandomName("APP")
+		proxyAppName = random_name.CATSRandomName("APP")
+		orgName = TestSetup.RegularUserContext().Org
+		spaceName = TestSetup.RegularUserContext().Space
+	})
+
+	AfterEach(func() {
+		app_helpers.AppReport(appName)
+		app_helpers.AppReport(proxyAppName)
+		Eventually(cf.Cf("delete", appName, "-f")).Should(Exit(0))
+		Eventually(cf.Cf("delete", proxyAppName, "-f")).Should(Exit(0))
+	})
+
+	Describe("when the readiness healthcheck is set to http", func() {
+		It("registers the route only when the readiness check passes", func() {
+			By("pushing the app")
+
+			Expect(cf.Cf("push", appName,
+				"-p", assets.NewAssets().Dora,
+				"-f", assets.NewAssets().Dora+"/readiness_manifest.yml",
+			).Wait(Config.CfPushTimeoutDuration())).To(Exit(0))
+
+			By("verifying the app starts")
+			Eventually(func() string {
+				return helpers.CurlAppRoot(Config, appName)
+			}, Config.DefaultTimeoutDuration()).Should(ContainSubstring("Hi, I'm Dora!"))
+
+			// Get the initial overlay IP for dora app
+			overlayIP := helpers.CurlApp(Config, appName, "/myip")
+
+			By("verifying the app is marked as ready")
+			Eventually(func() string {
+				return helpers.CurlApp(Config, appName, "/ready")
+			}, Config.DefaultTimeoutDuration()).Should(ContainSubstring("200 - ready"))
+
+			// TODO: only include this when audit events are built
+			// Eventually(func() string {
+			// 	return string(cf.Cf("events", appName).Wait().Out.Contents())
+			// }).Should(MatchRegexp("app.ready"))
+
+			Expect(string(logs.Recent(appName).Wait().Out.Contents())).Should(ContainSubstring("Container passed the readiness health check"))
+
+			By("triggering the app to make the /ready endpoint fail")
+			helpers.CurlApp(Config, appName, "/ready/false")
+
+			By("verifying the app is marked as not ready")
+
+			// TODO: only include this when audit events are built
+			// Eventually(func() string {
+			// 	return string(cf.Cf("events", appName).Wait().Out.Contents())
+			// }).Should(MatchRegexp("app.notready"))
+
+			Eventually(func() string {
+				return string(logs.Recent(appName).Wait().Out.Contents())
+			}, readinessHealthCheckTimeout).Should(ContainSubstring("Container failed the readiness health check"))
+
+			By("verifying the app is removed from the routing table")
+			Eventually(func() string {
```

Comment:
Should this be `Consistently` instead of `Eventually`?

Reply:
No, this was eventually to ensure that the delay between catching the readiness failure + route-emitter updating nats + gorouter. 

Arguably we could have another consistently check for the same thing after, but that would likely need to wait 40s+.

Reply:
I guess I expected the "Container failed the readiness health check" log message, or some comparable log message, to coincide with the app being removed from the routing table. Then after receiving that, using `Consistently` would make sense to check that the routing consistently doesn't work.

Reply:
> Container failed the readiness health check" log message, or some comparable log message, to coincide with the app being removed from the routing table. 

Sadly, this is not true, but I can see how this would be the assumption, and it is definitely confusing.

It looks something like this....
1. The readiness check gets a non-2xx. Marks the app as not-ready. Logs the not-ready bit in the logs.
2. Eventually (every 20s), the route emitter syncs, sees that the app is not ready and removes it from the routing table.

That's why it is Eventually and not Consistently.